prov/shm/src/smr_dsa.c

```diff
@@ -0,0 +1,848 @@
+			"doing cleanup\n");
+
+	for (i = 0; i < dsa_context->wq_count; i++)
+		dsa_idxd_wq_unmap(dsa_context->wq_portal[i]);
+
+	free(ep->dsa_context);
+}
+
+void smr_dsa_progress(struct smr_ep *ep)
+{
+	int index;
+	struct dsa_cmd_context *dsa_cmd_context;
+	uint8_t dsa_cmd_state;
+	struct smr_dsa_context *dsa_context = ep->dsa_context;
+
+	if (!dsa_is_work_in_progress(ep->dsa_context))
+		return;
+
+	pthread_spin_lock(&ep->region->lock);
+	for (index = 0; index < CMD_CONTEXT_COUNT; index++) {
+		dsa_cmd_context = dsa_get_cmd_context(dsa_context, index);
+
+		if (!dsa_cmd_context)
+			continue;
+
+		dsa_cmd_state =
+		    dsa_check_cmd_status(dsa_context, dsa_cmd_context);
+
+		if (dsa_cmd_state == DSA_CMD_COMPLETED)
+			dsa_process_complete_work(ep->region, dsa_cmd_context,
+					      dsa_context);
+	}
+	// Always signal the self to complete dsa, tx or rx.
+	smr_signal(ep->region);
+	pthread_spin_unlock(&ep->region->lock);
+}
+
+size_t smr_dsa_copy_to_sar(struct smr_ep *ep, struct smr_sar_pool *sar_pool,
+		struct smr_resp *resp, struct smr_cmd *cmd,
+		const struct iovec *iov, size_t count, size_t *bytes_done,
+		void *entry_ptr)
+{
+	struct dsa_cmd_context *dsa_cmd_context;
+
+	assert(smr_env.use_dsa_in_sar);
+
+	if (resp->status != SMR_STATUS_SAR_FREE)
+		return FI_EAGAIN;
+
+	dsa_cmd_context = dsa_allocate_cmd_context(ep->dsa_context);
+	if (!dsa_cmd_context)
+		return -FI_ENOMEM;
+
+	dsa_cmd_context->dir = OFI_COPY_IOV_TO_BUF;
+	dsa_cmd_context->entry_ptr = entry_ptr;
+	smr_dsa_copy_sar(sar_pool, ep->dsa_context, dsa_cmd_context, resp,
+			 cmd, iov, count, bytes_done, ep->region);
+
+	return FI_SUCCESS;
+}
+
+size_t smr_dsa_copy_from_sar(struct smr_ep *ep, struct smr_sar_pool *sar_pool,
+		struct smr_resp *resp, struct smr_cmd *cmd,
+		const struct iovec *iov, size_t count, size_t *bytes_done,
+		void *entry_ptr)
+{
+	struct dsa_cmd_context *dsa_cmd_context;
+
+	assert(smr_env.use_dsa_in_sar);
+
+	if (resp->status != SMR_STATUS_SAR_READY)
+		return FI_EAGAIN;
+
+	dsa_cmd_context = dsa_allocate_cmd_context(ep->dsa_context);
+	if (!dsa_cmd_context)
+		return -FI_ENOMEM;
+
+	dsa_cmd_context->dir = OFI_COPY_BUF_TO_IOV;
+	dsa_cmd_context->entry_ptr = entry_ptr;
+	smr_dsa_copy_sar(sar_pool, ep->dsa_context, dsa_cmd_context, resp,
+			 cmd, iov, count, bytes_done, ep->region);
+
+	return FI_SUCCESS;
+}
+
+#else
+
+size_t smr_dsa_copy_to_sar(struct smr_ep *ep, struct smr_sar_pool *sar_pool,
+		struct smr_resp *resp, struct smr_cmd *cmd,
+		const struct iovec *iov, size_t count, size_t *bytes_done,
+		void *entry_ptr)
+{
+	return -FI_ENOSYS;
+}
+
+size_t smr_dsa_copy_from_sar(struct smr_ep *ep, struct smr_sar_pool *sar_pool,
+		struct smr_resp *resp, struct smr_cmd *cmd,
+		const struct iovec *iov, size_t count, size_t *bytes_done,
+		void *entry_ptr)
+{
+	return -FI_ENOSYS;
+}
+
+void smr_dsa_context_init(struct smr_ep *ep)
+{
+	smr_env.use_dsa_in_sar = 0;
+	return;
+}
+
+void smr_dsa_context_cleanup(struct smr_ep *ep) {}
```

Comment:
Not sure @shefty has any opinions/nitpickiness on the structure of empty functions. I know in other places we have the normal style with a comment inside that says //noop